DS-3921 Now an item from workflow overview can be deleted

santit96 opened the pull request:
The problem was that sending back the item to the submitter and right after that delete the workspace item, generates an inconsistency
There was no need in sending the item back to submitter, as the item was going to be deleted anyway. Now the workflow item is directly removed

The whole issue is described here: https://jira.duraspace.org/browse/DS-3921

Fixes #7268

santit96 committed:
DS-3921 Now an item from workflow overview can be deleted

The problem was that sending back the item to the submitter and right after that delete the workspace item, generates an inconsistency
There was no need in sending the item back to submitter, as the item was going to be deleted anyway. Now the workflow item is directly removed

tdonohue commented:
The code here looks reasonable, but I have a general point of discussion to bring up here (not just for @santit96 but for everyone watching/looking at this ticket).

Should deletion of a workflow item send a notification back to the submitter?

The original code here calls `xmlWorkflowService.sendWorkflowItemBackSubmission()`, which performs two other tasks:
1. Adds a provenance entry noting the submission as rejected
2. Sends a notification back to the submitter that the submission is rejected.

The proposed change in this PR replaces that action with a simple deletion.  I agree that adding a provenance entry here has no purpose (if you are deleting the workflowitem anyhow).  But, do we need to ensure the submitter is still notified of this deletion?

(Admittedly, I don't have a strong opinion here either way....I just want to point out that while this PR will solve the bug, it also will change the behavior slightly.)

mwoodiupui commented:
How, then, will the submitter learn that the submission was deleted?  Were I the submitter, I would want the notification.

cjuergen commented:
There should be a notification, but more useful would be an in workflow communication. 
Working with submissions by faculty members on a daily basis, one often needs to clear something (mostly rights/licenses) without rejecting the submission. 
One can communicate directly with the submitter, but this is "off" the workflow and the information not available for other members of the reviewing process.

tdonohue commented:
Based on the discussion in this ticket (see above), it sounds like we need to recommend this PR be refactored to still include a notification to the submitter.  This notification should be similar to what is provided in [`XMLWorkflowService.sendWorkflowItemBackSubmission`](https://github.com/DSpace/DSpace/blob/dspace-6_x/dspace-api/src/main/java/org/dspace/xmlworkflow/XmlWorkflowServiceImpl.java#L754), but it could even be a new method in `XMLWorkflowService` which simply sends a deletion notification & deletes the WorkflowItem (perhaps a new `XMLWorkflowService.sendWorkflowItemDelete()` method needs to exist)

I'll also add these notes to the ticket itself.  For now, I'm flagging this as "work in progress" to signify that we won't look to merge this until further work is complete.  Please let us know @santit96 if you are willing to still work on this, or if we should look for a different volunteer.  Thanks!

santit96 commented:
Thanks for the feedback, i will work on it

santit96 committed:
DS-3921 A notitfication to the submitter has been added when an administrator deletes an item

Now when an administrator deletes an item from the workflowoverview view the system sends a notification to the submitter
There is a new method in XMLWorkflowService which deletes the workflowitem and sends a notification, as an email, to the user who submitted the item

santit96 commented:
As @tdonohue suggested, i created a new XMLWorkflowService.sendWorkflowItemDelete() method which deletes the workflowitem and sends a notification to the submitter

tdonohue commented:
As this changes the behavior of XMLWorkflow, I'd recommend we recreate this PR for 7.x (`main` branch).  6.x is bug fix only at this time.

Also, it's worth noting that WorkflowItems can be deleted in 7.0 by an Administrator.  So, this feature exists for Administrators in 7.0
